### ISSUE TITLE
compiler generated fields

### ISSUE THREAD

**Issue by GotoFinal** [0xd4d/dnSpy#694]
After decompiling in many places you can still see some generated stuff like this:
```csharp
        [CompilerGenerated]
        private static ParameterizedThreadStart <>f__mg$cache0;
```
and usage of it:
```csharp
            if (HTTPCacheService.<>f__mg$cache0 == null)
            {
                HTTPCacheService.<>f__mg$cache0 = new ParameterizedThreadStart(HTTPCacheService.ClearImpl);
            }
            new Thread(HTTPCacheService.<>f__mg$cache0).Start();
```
But it should be decompiled to much simpler structure:
```csharp
new Thread(new ParameterizedThreadStart(HTTPCacheService.ClearImpl)).Start();
```

**Comment by GotoFinal** [comment]
Also if anyone need fast workaround, you can use this regex and hope that it will not break anything:  
`if \(([A-Za-z0-9]+\.\<\>[a-z_]+\$cache[0-9a-fA-Z]+) == null\)\s*\{\s*\s\1 = (.*);\s*\}\s*(.+?)(\1)(.+)`  
replace to:  
`$3$2$5`  

And then replace this with nothing:  
`\[CompilerGenerated\]\s*(?:.*?)\$cache[A-Za-z0-9]+\;\s*`

**Comment by Hunteristic** [comment]
Found the workaround extremely useful, thank you very much.

**Comment by pfgithub** [comment]
How do you deal with the generated fields like `private static ParameterizedThreadStart <>f__mg$cache0;`?

**Comment by kusalux** [comment]
<details><summary>Spoiler</summary>
<p>

> [global::System.Runtime.CompilerServices.CompilerGenerated]
> private static global::System.Func<string, bool> <>f__mg$cache0;
> 
> [global::System.Runtime.CompilerServices.CompilerGenerated]
> private static global::UnityEngine.Events.UnityAction <>f__mg$cache1;
> 
> [global::System.Runtime.CompilerServices.CompilerGenerated]
> private static global::UnityEngine.Events.UnityAction<global::PW.Networking.PaymentResponseCode, global::PW.Networking.NutakuBillingItem> <>f__mg$cache2;
> 
> [global::System.Diagnostics.DebuggerBrowsable(global::System.Diagnostics.DebuggerBrowsableState.Never)]
> public static event global::PW.Networking.AchievementHandler OnAchievementUnlocked;
> [global::System.Diagnostics.DebuggerBrowsable(global::System.Diagnostics.DebuggerBrowsableState.Never)]
> public static event global::PW.Networking.AchievementHandler OnAchievementPrizeCollected;

</p>
</details>
I did have other problematic compiled code too

> 

Thus created regex:
`(\[global[^\s\[\]]+(?:CompilerGenerated|DebuggerBrowsableState\.Never\))\])(?! *get| *set)([^${}\[]*(?:\$cache[0-9a-fA-Z]+;))?\s*`

If you don't want all (supposed non-problematic) CompilerGenerated entries to be deleted ... for ease of use run once with last "?" removed, once with "CompilerGenerated|" removed.

**Comment by StijnSimons** [comment]
I tried the regex, but it got no matches? Any idea whats the problem?

(One of the broken syntaxes)
```csharp
Logger.OnLogMessage -= InControlManager.<>f__mg$cache0;
if (InControlManager.<>f__mg$cache1 == null)
{
    InControlManager.<>f__mg$cache1 = new Action<LogMessage>(InControlManager.LogMessage);
}
Logger.OnLogMessage += InControlManager.<>f__mg$cache1;
```

**Comment by StijnSimons** [comment]
Bump?

**Comment by StijnSimons** [comment]
I used both regexes (they work now for some reason) but i'm still stuck with these ones:
```
this.demoHub.ComplexArray(new object[]
			{
				<>__AnonType,
				<>__AnonType,
				<>__AnonType
			});
```

**Comment by kusalux** [comment]
Did you try with other decompilers like JustDecompile? So you know, what it should look like? Then creating a regex out to fix it is rather easy.

**Comment by StijnSimons** [comment]
Yes, i have tried using JustDecompile, but it ended up with an even worse result. I’m not familliar with regular expressions so i don’t know if i can do that.

**Comment by smygarn** [comment]
I would prefer to not generate the code at all instead of generating code that doesn't compile?

My output is spammed by <>f__mg$cache0 .. <>f__mg$cache9

**Comment by MapleWheels** [comment]
Those types and names are completely valid from the CLR's point of view, but those types of symbols are restricted to compiler use only, so VS is going to complain when it sees them.
